File: Console/Classes/ApplicationDelegate.m
/*******************************************************************************
 * The MIT License (MIT)
 * 
 * Copyright (c) 2015 Jean-David Gadina - www-xs-labs.com
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 ******************************************************************************/

/*!
 * @file        ApplicationDelegate.m
 * @copyright   (c) 2016, Jean-David Gadina - www.xs-labs.com
 */

#import "ApplicationDelegate.h"
#import "Console-Swift.h"
#import "MainWindowController.h"
#import "Preferences.h"
#import "PreferencesWindowController.h"

@interface ApplicationDelegate()

@property( atomic, readwrite, strong ) NSMutableArray              * mainWindowControllers;
@property( atomic, readwrite, strong ) AboutWindowController       * aboutWindowController;
@property( atomic, readwrite, strong ) PreferencesWindowController * preferencesWindowController;

- ( IBAction )newDocument: ( id )sender;

@end

@implementation ApplicationDelegate

- ( void )applicationDidFinishLaunching: ( NSNotification * )notification
{
    ( void )notification;
    
    self.mainWindowControllers = [ NSMutableArray new ];
    
    [ self newDocument: nil ];
    [ Preferences sharedInstance ].lastStart = [ NSDate date ];
}

- ( void )applicationWillTerminate: ( NSNotification * )notification
{
    ( void )notification;
}

- ( BOOL )applicationShouldTerminateAfterLastWindowClosed: ( NSApplication * )sender
{
    ( void )sender;
    
    return NO;
}

- ( IBAction )newDocument: ( nullable id )sender
{
    MainWindowController * controller;
    
    ( void )sender;
    
    controller = [ MainWindowController new ];
    
    if( [ Preferences sharedInstance ].lastStart == nil )
    {
        [ controller.window center ];
    }
    
    [ self.mainWindowControllers addObject: controller ];
    [ controller.window makeKeyAndOrderFront: nil ];
}

- ( IBAction )showAboutWindow: ( id )sender
{
    if( self.aboutWindowController == nil )
    {
        self.aboutWindowController = [ AboutWindowController new ];
        
        [ self.aboutWindowController.window center ];
    }
    
    [ self.aboutWindowController.window makeKeyAndOrderFront: sender ];
}

- ( IBAction )showPreferencesWindow: ( id )sender
{
    if( self.preferencesWindowController == nil )
    {
        self.preferencesWindowController = [ PreferencesWindowController new ];
        
        [ self.preferencesWindowController.window center ];
    }
    
    [ self.preferencesWindowController.window makeKeyAndOrderFront: sender ];
}

@end

File: Console/Classes/ApplicationDelegate.swift
/*******************************************************************************
 * The MIT License (MIT)
 * 
 * Copyright (c) 2017 Jean-David Gadina - www.xs-labs.com
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 ******************************************************************************/

import Cocoa

@NSApplicationMain @objc public class ApplicationDelegate: NSObject, NSApplicationDelegate
{
    @objc private dynamic var aboutWindowController:       AboutWindowController?
    @objc private dynamic var preferencesWindowController: PreferencesWindowController?
    @objc private dynamic var mainWindowControllers:       [ MainWindowController ] = []
    
    @objc public func applicationDidFinishLaunching( _ notification: Notification )
    {
        NotificationCenter.default.addObserver( self, selector: #selector( windowWillClose ), name: NSWindow.willCloseNotification, object: nil )
        self.newDocument( nil )
        
        Preferences.shared.lastStart = Date()
    }
    
    @objc public func applicationWillTerminate( _ notification: Notification )
    {
        NotificationCenter.default.removeObserver( self )
    }
    
    @objc public func applicationShouldTerminateAfterLastWindowClosed( _ sender: NSApplication ) -> Bool
    {
        return false
    }
    
    @IBAction func showAboutWindow( _ sender: Any? )
    {
        if( self.aboutWindowController == nil )
        {
            self.aboutWindowController = AboutWindowController()
            
            self.aboutWindowController?.window?.center()
        }
        
        self.aboutWindowController?.window?.makeKeyAndOrderFront( sender )
    }
    
    @IBAction func showPreferencesWindow( _ sender: Any? )
    {
        if( self.preferencesWindowController == nil )
        {
            self.preferencesWindowController = PreferencesWindowController()
            
            self.preferencesWindowController?.window?.center()
        }
        
        self.preferencesWindowController?.window?.makeKeyAndOrderFront( sender )
    }
    
    @IBAction func newDocument( _ sender: Any? )
    {
        let controller = MainWindowController()
        
        if( Preferences.shared.lastStart == nil )
        {
            controller.window?.center()
        }
        
        self.mainWindowControllers.append( controller )
        controller.window?.makeKeyAndOrderFront( sender )
    }
    
    @objc public func windowWillClose( _ notification: Notification )
    {
        guard let window = ( notification.object as AnyObject? ) as? NSWindow else
        {
            return
        }
        
        if( window == self.preferencesWindowController?.window )
        {
            self.preferencesWindowController = nil
        }
        
        self.mainWindowControllers = self.mainWindowControllers.filter{ $0 != window.windowController }
    }
}
